GeniusVentures/wallet-core
# Windows build and set output location of rust library

## Description

Incremented the project by adding Windows MSVC build support. Added configuration to either shared or static library and if in case of static the visibility of the exported symbols will be set to default to make it visible to dart ffi, which requires indirect loading by its own shared library wrappers. Fixed the output path of the rust library to the same used by cmake. The protobuf and protoc libraries are searched for in case the external protobuf project doesn't provide the same variables the internal protobuf does, this was done also to solve the naming differences between Unix and Windows on libraries.
Some minor changes were made in the source code to adapt do MSVC, mostly considering some casts from iterators are not allowed. The functionality of static libraries with "default" visibility symbols were tested in OSX, iOS, Android and Linux on dart ffi, but are not tested in Windows still.

The functionality when calling the scripts as the original project did was not tested. 
The xcode generation is not working with external protobuf, due to dependencies of the project to both the revision used by it and the paths set in symbolic links.

## How to test

For Windows build instead of generating the files using "tools/generate-files.sh" one needs to open a "cmd.exe" and call "tools/win-generate-files.bat" with the same arguments expected. Ex:

```
win-generate-files.bat --target-os=windows --protobuf-dir=C:/temp/my_grpc --out-dir=C:/temp/wallet-core-out --plugin-dir=C:/temp/protobuf-plugin-out/bin
```
The script works with either '\' or '/' on the path naming. Spaces on the folders weren't tested.

The output of this script will generate files inside the "wallet-core" project folder, as well as the "wallet_core_rs.lib" file on the "C:/temp/wallet-core-out/lib" directory. After that the compiling of the project can be done using cmake.

In order to test the type of library configuration, one can define the cmake variable "TW_STATIC_LIBRARY" to TRUE to make it compile to static, otherwise it will generate a shared one with symbols hidden by default.

## Types of changes

* New feature: Windows MSVC build supported.
* New feature: Configuration of rust library to either Windows or Unix systems.
* New feature: Type of library configurable (static or shared)
* New feature: Visibility set to default if the library is being statically compiled.
* Bug fix: xcodegen call only used if "ios" build selected and the user informs the parameter "-x" on the generate-files.sh script.
* Bug fix: Protobuf and protoc libraries are searched for if not defined. This also works for Windows or Unix.


## Checklist
